adapters/celestia/src/verifier/address.rs

```diff
@@ -1,64 +1,116 @@
 use std::fmt::{Display, Formatter};
 use std::str::FromStr;
 
+use bech32::WriteBase32;
 use borsh::{BorshDeserialize, BorshSerialize};
 use serde::{Deserialize, Serialize};
 use sov_rollup_interface::AddressTrait;
 
+/// Human Readable Part: "celestia" for Celestia network
 const HRP: &str = "celestia";
 
+/// Representation of the address in the Celestia network
 #[derive(Debug, PartialEq, Clone, Eq, Serialize, Deserialize, BorshDeserialize, BorshSerialize)]
-// Raw ASCII bytes, including HRP
-// TODO: https://github.com/Sovereign-Labs/sovereign-sdk/issues/469
-pub struct CelestiaAddress(Vec<u8>);
+pub struct CelestiaAddress {
+    // Storing vector of u5 each as u8
+    data: Vec<u8>,
+    // indication if it is Bech32m or not
+    m: bool,
+}
+
+impl CelestiaAddress {
+    fn variant(&self) -> bech32::Variant {
+        if self.m {
+            bech32::Variant::Bech32m
+        } else {
+            bech32::Variant::Bech32
+        }
+    }
+}
 
 impl AsRef<[u8]> for CelestiaAddress {
     fn as_ref(&self) -> &[u8] {
-        self.0.as_ref()
+        self.data.as_ref()
     }
 }
 
+/// Decodes slice of bytes into CelestiaAddress
+/// Treats it as string if it starts with HRP and the rest is valid ASCII
+/// Otherwise just checks if it contains valid `u5` elements.
+/// Note: Assumes that passed bytes is Bech32m encoding, not Bech32!
 impl<'a> TryFrom<&'a [u8]> for CelestiaAddress {
     type Error = anyhow::Error;
 
     fn try_from(value: &'a [u8]) -> Result<Self, Self::Error> {
-        Ok(Self(value.to_vec()))
+        if value.len() >= HRP.len() && &value[..HRP.len()] == HRP.as_bytes() && value.is_ascii() {
+            // safety, because we checked that it is ASCII
+            let s = unsafe { std::str::from_utf8_unchecked(value) };
+            return CelestiaAddress::from_str(s);
+        }
+        let mut data = Vec::with_capacity(value.len());
+        for (idx, &item) in value.iter().enumerate() {
+            bech32::u5::try_from_u8(item)
+                .map_err(|e| anyhow::anyhow!("Element at {} is not u5: {}", idx, e))?;
+            data.push(item);
```

Comment:
Going along with my previous comment, can we use a [u8;32] internal representation and just use the `bech32::decode` function here?

Reply:
`bech32::decode` has following syntax:

```rust
pub fn decode(s: &str) -> Result<(String, Vec<u5>, Variant), Error>
```

Which decodes string. 
we store array of 32 u5s internally, so we should make sure, that each element actually valid u5.

Anyway, I believe that according to https://github.com/Sovereign-Labs/sovereign-sdk/issues/493 it is better to remove it and always use `TryFrom<[u8]>` so errors can be handled